### PR TITLE
IMPROVEMENT] Refactor Typed Signature

### PR DESCRIPTION
**Pull request by blackdevelopa**
**Development & PR Process**
1. Follow MetaMask [Mobile Coding Standards](https://docs.google.com/document/d/1VJLwTRsUw_5EDq_o8d6sSbXUAYENLurkRitYO45eY5o/edit?usp=sharing)
2. Add `release-xx` label to identify the PR slated for a upcoming release (will be used in release discussion)
3. Add `needs-dev-review` label when work is completed
4. Add `needs-qa` label when dev review is completed
5. Add `QA Passed` label when QA has signed off

**Description**
The PR refactors the TypedSign Signature code by doing two things. First, convert TypedSign code to TypeScript and moved repeated signature code to hooks and util (SignatureRequest UI).

It shouldn't affect the current behaviour.

**Screenshots/Recordings**

_If applicable, add screenshots and/or recordings to visualize the before and after of your change_

**Issue**

Progresses #???

**Checklist**

* [x] There is a related GitHub issue
* [ ] Tests are included if applicable
* [ ] Any added code is fully documented

### CODE REVIEW

**Comment by jpuri** [comment]
Hey @blackdevelopa : for refactoring PRs I think we should have a well defined github issue and an agreed upon approach.

File: app/components/UI/SignatureRequest/SignatureSharedState/signatureAnalytics.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+      sign_type: type,
+      version: messageParams?.version,
+      ...currentPageInformation?.analytics,
+    };
+  } catch (error) {
+    return {};
+  }
+};
+
+export default getSignatureAnalytics;
```
Comment (jpuri) [inline app/components/UI/SignatureRequest/SignatureSharedState/signatureAnalytics.ts:29]: @blackdevelopa : getSignatureAnalytics is a confusing name.

An option here is we create a hook and move this and even more of analytics related code to it.

File: app/components/UI/SignatureRequest/Util/showWCNotification.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        duration: 5000,
+        title: confirmation
+          ? strings('notifications.wc_signed_title')
+          : strings('notifications.wc_signed_rejected_title'),
+        description: strings('notifications.wc_description'),
+      });
+  });
+};
+
+export default showWalletConnectNotification;
```
Comment (jpuri) [inline app/components/UI/SignatureRequest/Util/showWCNotification.ts:29]: 👍

File: app/components/UI/SignatureRequest/Util/signatureAnalytics.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+      sign_type: type,
+      version: messageParams?.version,
+      ...currentPageInformation?.analytics,
+    };
+  } catch (error) {
+    return {};
+  }
+};
+
+export default getSignatureAnalytics;
```
Comment (jpuri) [inline app/components/UI/SignatureRequest/Util/signatureAnalytics.ts:29]: How is this different: https://github.com/MetaMask/metamask-mobile/pull/5629/files#diff-aa4a18d3689d7a8c330802ed4207a46a4c3adc7035519571d5d9a52e49296fb9

File: app/components/hooks/Signatures/useMessage.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+      confirmation: false,
+      messageParams,
+      isError: false,
+    });
+  };
+
+  return [rejectMessage, signMessage];
+};
+
+export default useMessage;
```
Comment (jpuri) [inline app/components/hooks/Signatures/useMessage.ts:50]: name `useMessage` not very intuitive

File: app/components/UI/SignatureRequest/SignatureSharedState/signatureAnalytics.ts (hunk 1 above)
Comment (blackdevelopa) [reply app/components/UI/SignatureRequest/SignatureSharedState/signatureAnalytics.ts]: Good catch. Removed. https://github.com/MetaMask/metamask-mobile/pull/5629/commits/a9c0cdf886248a68f8965ed9b8ddb11799962691

File: app/components/UI/SignatureRequest/Util/signatureAnalytics.ts (hunk 3 above)
Comment (blackdevelopa) [reply app/components/UI/SignatureRequest/Util/signatureAnalytics.ts]: Updated. https://github.com/MetaMask/metamask-mobile/pull/5629/commits/db7b3b720ac090d2f0fe8f2e87f4a5e1932659d4

File: app/components/UI/SignatureRequest/SignatureSharedState/signatureAnalytics.ts (hunk 1 above)
Comment (blackdevelopa) [reply app/components/UI/SignatureRequest/SignatureSharedState/signatureAnalytics.ts]: Removed. https://github.com/MetaMask/metamask-mobile/pull/5629/commits/a9c0cdf886248a68f8965ed9b8ddb11799962691

File: app/components/hooks/Signatures/useMessage.ts (hunk 4 above)
Comment (blackdevelopa) [reply app/components/hooks/Signatures/useMessage.ts]: This makes sense. Updated now. https://github.com/MetaMask/metamask-mobile/pull/5629/commits/a9c0cdf886248a68f8965ed9b8ddb11799962691

File: app/components/hooks/Signatures/useSignatureTypes.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  /**
+   * Indicated whether or not the expanded message is shown
+   */
+  showExpandedMessage?: boolean;
+}
+
+export interface UseMessageProps {
+  messageParams: any;
+  type: string;
+}
```
[the remark this answers is not in the document]
Comment (blackdevelopa) [reply app/components/hooks/Signatures/useSignatureTypes.ts]: Updated too. https://github.com/MetaMask/metamask-mobile/pull/5629/commits/a9c0cdf886248a68f8965ed9b8ddb11799962691

**Comment by blackdevelopa** [comment]
Hey @jpuri, sure. We can pair program on this when you have some time. [Link](https://docs.google.com/document/d/181J1OwvZ9mWQykpjOl_avULY8clFYhEhaiE1Oq29MmA/edit?usp=sharing)

File: app/components/UI/SignatureRequest/Util/index.ts [hunk 6]
```diff
@@ -0,0 +1,2 @@
+export { default as signatureAnalytics } from './signatureAnalytics';
+export { default as showWalletConnectNotification } from './showWCNotification';
```
Comment (jpuri) [inline app/components/UI/SignatureRequest/Util/index.ts:2]: I think we can create a single `util.ts` also as I do not expect it to grow a lot and I find it convenient to have all util methods in one file.

File: app/components/UI/SignatureRequest/Util/signatureAnalytics.ts [hunk 7]
```diff
@@ -0,0 +1,29 @@
+import Engine from '../../../../core/Engine';
+import { getAddressAccountType } from '../../../../util/address';
+import { SignatureAnalayticsProps } from '../types';
+
```
Comment (jpuri) [inline app/components/UI/SignatureRequest/Util/signatureAnalytics.ts:4]: Method below can be named better.

File: app/components/UI/TypedSign/TypedSign.tsx [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+
+  const signatureAnalyticsCallback = useCallback(
+    () =>
+      signatureAnalytics({
+        currentPageInformation,
+        type: SignatureType.TYPED,
+        messageParams,
+      }),
+    [currentPageInformation, messageParams],
+  );
```
Comment (jpuri) [inline app/components/UI/TypedSign/TypedSign.tsx:49]: Above we are not gaining anything by using `useCallback` as the function declaration does not change with state.

File: app/components/UI/TypedSign/TypedSign.tsx [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+    if (truncateMessage) return;
+    if (
+      (Device.isIos() && e.nativeEvent.layout.height > 70) ||
+      (Device.isAndroid() && e.nativeEvent.layout.height > 100)
+    ) {
+      setTruncateMessage(true);
+      return;
+    }
+    setTruncateMessage(false);
+  };
```
Comment (jpuri) [inline app/components/UI/TypedSign/TypedSign.tsx:96]: This logic can go to util

```
if (
      (Device.isIos() && e.nativeEvent.layout.height > 70) ||
      (Device.isAndroid() && e.nativeEvent.layout.height > 100)
    )
  ```

File: app/components/hooks/Signatures/useSignatureAction.ts [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+      confirmation: false,
+      messageParams,
+      isError: false,
+    });
+  };
+
+  return [rejectMessage, signMessage];
+};
+
+export default useSignatureAction;
```
Comment (jpuri) [inline app/components/hooks/Signatures/useSignatureAction.ts:55]: It will be nice to move analytics code also to this hook.

**Pull request closed without merging**